src/Eshopworld.Messaging/Messenger.cs

```diff
@@ -1,145 +1,196 @@
+    using System.Reactive.Linq;
+    using System.Reactive.Subjects;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// The main messenger entry point.
+    /// It is recommended that only one messenger exists per application so you should always give this a singleton lifecycle.
+    /// </summary>
+    /// <remarks>
+    /// Some checks are only valuable if you have a unique messenger class in your application. If you set it up as transient
+    /// then some checks will only run against a specific instance and won't really validate the entire behaviour.
+    /// </remarks>
+    public class Messenger : IDoFullMessaging, IDoFullReactiveMessaging
+    {
+        internal readonly object Gate = new object();
+        internal readonly string ConnectionString;
+        internal readonly string SubscriptionId;
+
+        internal readonly ISubject<object> MessagesIn = new Subject<object>();
+
+        internal Dictionary<Type, IDisposable> MessageSubs = new Dictionary<Type, IDisposable>();
+        internal Dictionary<Type, ServiceBusAdapter> ServiceBusAdapters = new Dictionary<Type, ServiceBusAdapter>();
+
+        internal ServiceBusAdapter<T> GetQueueAdapterIfExists<T>() where T : class =>
+            ServiceBusAdapters.TryGetValue(typeof(T), out var result)
+                ? (ServiceBusAdapter<T>)result
+                : throw new InvalidOperationException($"Messages/events of type {typeof(T).FullName} haven't been setup properly yet");
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="Messenger"/>.
+        /// </summary>
+        /// <param name="connectionString">The Azure Service Bus connection string.</param>
+        /// <param name="subscriptionId">The subscription ID where the service bus namespace lives.</param>
+        public Messenger([NotNull]string connectionString, [NotNull]string subscriptionId)
+        {
+            ConnectionString = connectionString;
+            SubscriptionId = subscriptionId;
+        }
+
+        /// <inheritdoc />
+        public async Task Send<T>(T message)
+            where T : class
+        {
+            if (!ServiceBusAdapters.ContainsKey(typeof(T)))
+            {
+                SetupMessageType<T>(10, MessagingTransport.Queue);
+            }
+
+            await ((QueueAdapter<T>)ServiceBusAdapters[typeof(T)]).Send(message).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public async Task Publish<T>(T @event)
+            where T : class
+        {
+            if (!ServiceBusAdapters.ContainsKey(typeof(T)))
+            {
+                SetupMessageType<T>(10, MessagingTransport.Topic);
+            }
+
+            await ((TopicAdapter<T>)ServiceBusAdapters[typeof(T)]).Send(@event).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void Receive<T>(Action<T> callback, int batchSize = 10)
+            where T : class
+        {
+            if (MessageSubs.ContainsKey(typeof(T)))

```

Comment:
locking?

Reply:
As @dguerin pointed out, Setup has a lock inside it. There is one possible race condition here where you can dual subscribe _sometimes_, but that will be picked up.

Reply:
AFAIK ContainsKey is not thread save when the dictionary is modified, so locking in Setup won't help.

Reply:
Good point @arturek, I'll swap everything to `ConcurrentDictionary`, which was what we did with Telemetry to kill all the possible concurrency racing conditions!